# Analyzing borrowers’ risk of defaulting

Your project is to prepare a report for a bank’s loan division. You’ll need to find out if a customer’s marital status and number of children has an impact on whether they will default on a loan. The bank already has some data on customers’ credit worthiness.

Your report will be considered when building a **credit scoring** of a potential customer. A ** credit scoring ** is used to evaluate the ability of a potential borrower to repay their loan.

## Open the data file and have a look at the general information. 

In [1]:
import pandas as pd
#Read csv file
data = pd.read_csv('https://code.s3.yandex.net/datasets/credit_scoring_eng.csv') 
#General information
data.info()
data

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


children  days_employed  dob_years            education  education_id  \
0             1   -8437.673028         42    bachelor's degree             0   
1             1   -4024.803754         36  secondary education             1   
2             0   -5623.422610         33  Secondary Education             1   
3             3   -4124.747207         32  secondary education             1   
4             0  340266.072047         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21520         1   -4529.316663         43  secondary education             1   
21521         0  343937.404131         67  secondary education             1   
21522         1   -2113.346888         38  secondary education             1   
21523         3   -3112.481705         38  secondary education             1   
21524         2   -1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21520  civil partnership                 1      F    business     0   
21521            married                 0      F     retiree     0   
21522  civil partnership                 1      M    employee     1   
21523            married                 0      M    employee     1   
21524            married                 0      F    employee     0   

       total_income                  purpose  
0         40620.102    purchase of the house  
1         17932.802             car purchase  
2         23341.752    purchase of the house  
3         42820.568  supplementary education  
4         25378.572        to have a wedding  
...             ...                      ...  
21520     35966.698     housing transactions  
21521     24959.969        purchase of a car  
21522     14347.610                 property  
21523     39054.888        buying my own car  
21524     13127.587             to buy a car  

[21525 rows x 12 columns]

### Conclusion

1. There are 2174 missing values at 'days_employed' column
2. There are 2174 missing values at 'total_income' column
3. The type of 'debt' column is int64 and should be boolean

## Data preprocessing

### Processing missing values

In [2]:
#Find out how many missing value for each column
data.isnull().sum()

#fill missing value in 'days_employed' column with 0
data['days_employed'].fillna(0, inplace=True)

In [3]:
#Check all rows were 'total_income' = NAN
data[data['total_income'].isnull()]

children  days_employed  dob_years            education  education_id  \
12            0            0.0         65  secondary education             1   
26            0            0.0         41  secondary education             1   
29            0            0.0         63  secondary education             1   
41            0            0.0         50  secondary education             1   
55            0            0.0         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            0.0         47  Secondary Education             1   
21495         1            0.0         50  secondary education             1   
21497         0            0.0         48    BACHELOR'S DEGREE             0   
21502         1            0.0         42  secondary education             1   
21510         2            0.0         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]

In [4]:
#customers with age=0
data[data['dob_years'] == 0]

children  days_employed  dob_years            education  education_id  \
99            0  346541.618895          0  Secondary Education             1   
149           0   -2664.273168          0  secondary education             1   
270           3   -1872.663186          0  secondary education             1   
578           0  397856.565013          0  secondary education             1   
1040          0   -1158.029561          0    bachelor's degree             0   
...         ...            ...        ...                  ...           ...   
19829         0       0.000000          0  secondary education             1   
20462         0  338734.868540          0  secondary education             1   
20577         0  331741.271455          0  secondary education             1   
21179         2    -108.967042          0    bachelor's degree             0   
21313         0   -1268.487728          0  secondary education             1   

      family_status  family_status_id gender income_type  debt  total_income  \
99          married                 0      F     retiree     0     11406.644   
149        divorced                 3      F    employee     0     11228.230   
270         married                 0      F    employee     0     16346.633   
578         married                 0      F     retiree     0     15619.310   
1040       divorced                 3      F    business     0     48639.062   
...             ...               ...    ...         ...   ...           ...   
19829       married                 0      F    employee     0           NaN   
20462       married                 0      F     retiree     0     41471.027   
20577     unmarried                 4      F     retiree     0     20766.202   
21179       married                 0      M    business     0     38512.321   
21313     unmarried                 4      M    employee     0     24387.070   

                            purpose  
99                              car  
149            housing transactions  
270              housing renovation  
578    construction of own property  
1040                   to own a car  
...                             ...  
19829                       housing  
20462      purchase of my own house  
20577                      property  
21179        building a real estate  
21313          housing transactions  

[101 rows x 12 columns]

In [5]:
#Check if there is a relation between customers with age=0 to repaying a loan on time
data[data['dob_years'] == 0].groupby('debt').count()

children  days_employed  dob_years  education  education_id  \
debt                                                                
0           93             93         93         93            93   
1            8              8          8          8             8   

      family_status  family_status_id  gender  income_type  total_income  \
debt                                                                       
0                93                93      93           93            83   
1                 8                 8       8            8             8   

      purpose  
debt           
0          93  
1           8

In [6]:
#Check how many customers with total_income=NAN are unemployed 'days_employed' = 0
uneployed_without_income =  data.loc[ (data['total_income'].isnull()) & (data['days_employed'] ==0 )]
#Get numbers of rows
uneployed_without_income.shape[0]

2174

In [7]:
# rows with days_employed<0
data[data['days_employed']< 0]

children  days_employed  dob_years            education  education_id  \
0             1   -8437.673028         42    bachelor's degree             0   
1             1   -4024.803754         36  secondary education             1   
2             0   -5623.422610         33  Secondary Education             1   
3             3   -4124.747207         32  secondary education             1   
5             0    -926.185831         27    bachelor's degree             0   
...         ...            ...        ...                  ...           ...   
21519         1   -2351.431934         37      graduate degree             4   
21520         1   -4529.316663         43  secondary education             1   
21522         1   -2113.346888         38  secondary education             1   
21523         3   -3112.481705         38  secondary education             1   
21524         2   -1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
5      civil partnership                 1      M    business     0   
...                  ...               ...    ...         ...   ...   
21519           divorced                 3      M    employee     0   
21520  civil partnership                 1      F    business     0   
21522  civil partnership                 1      M    employee     1   
21523            married                 0      M    employee     1   
21524            married                 0      F    employee     0   

       total_income                     purpose  
0         40620.102       purchase of the house  
1         17932.802                car purchase  
2         23341.752       purchase of the house  
3         42820.568     supplementary education  
5         40922.170       purchase of the house  
...             ...                         ...  
21519     18551.846  buy commercial real estate  
21520     35966.698        housing transactions  
21522     14347.610                    property  
21523     39054.888           buying my own car  
21524     13127.587                to buy a car  

[15906 rows x 12 columns]

In [8]:
#replace negative days_employed to positive
data.loc[data['days_employed']< 0, 'days_employed'] = data.loc[data['days_employed']< 0, 'days_employed'].abs()

In [9]:
# rows with days_employed<0
#verify no negative numbers left
data[data['days_employed']< 0]

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose]
Index: []

In [10]:
def new_days_employed(days_employed):
    #days_employed = row['days_employed']
    #21360 = 60 years. above this number divide by 10 until getting realistic number
   
    while days_employed > 21360:
        days_employed = days_employed /10
    return days_employed

data['days_employed'] = data['days_employed'].apply(new_days_employed)
        

In [11]:
#Conclusion - all customers with total_income=NAN are unemployed so there 'total_income' is 0
#fill missing value in 'total_income' column with 0
#data['total_income'].fillna(0, inplace=True)

In [12]:

#Calculate mean of  'total_income' column by 'income_type'
income_type_mean_value = data.groupby('income_type')['total_income'].mean()
#Calculate mean of  'days_employed' column by 'income_type'
days_employed_mean_value = data.groupby('income_type')['days_employed'].mean()

#Replace NaN and 0 values from  'total_income' and 'days_employed' with mean values
for index, row in data.iterrows():

    total_income = row['total_income']
    income_type = row['income_type']
    days_employed = row['days_employed']

    if (pd.isna(total_income)) | (total_income == 0):        
        data.at[index,'total_income'] =  income_type_mean_value[income_type]
       
    if (pd.isna(days_employed)) | (days_employed == 0): 
        data.at[index, 'days_employed'] =  days_employed_mean_value[income_type]





In [13]:
data.loc[data['total_income'].isnull()]['total_income']

Series([], Name: total_income, dtype: float64)

In [14]:
#Check children <0
data[data['children'] <0]

children  days_employed  dob_years            education  education_id  \
291          -1    4417.703588         46  secondary education             1   
705          -1     902.084528         50  secondary education             1   
742          -1    3174.456205         57  secondary education             1   
800          -1    3499.878522         54  secondary education             1   
941          -1    3259.094970         57  Secondary Education             1   
1363         -1    1195.264956         55  SECONDARY EDUCATION             1   
1929         -1    1461.303336         38  secondary education             1   
2073         -1    2539.761232         42  secondary education             1   
3814         -1    3045.290443         26  Secondary Education             1   
4201         -1     901.101738         41  secondary education             1   
4402         -1    3980.013029         64  SECONDARY EDUCATION             1   
4542         -1    1811.899756         32  secondary education             1   
5272         -1    1143.485347         46  secondary education             1   
6013         -1    1361.258696         46    bachelor's degree             0   
6385         -1    3702.154762         48  secondary education             1   
7190         -1    5928.202068         34  SECONDARY EDUCATION             1   
7278         -1     526.318451         51  secondary education             1   
7615         -1    2095.293025         35  secondary education             1   
7685         -1    3237.360455         53  secondary education             1   
8229         -1    1803.441074         54    bachelor's degree             0   
8251         -1    3113.998449         31  secondary education             1   
8561         -1     195.479496         31  secondary education             1   
9565         -1    2896.629686         34    bachelor's degree             0   
9574         -1    2710.419901         28    bachelor's degree             0   
10177        -1    1743.604011         37  secondary education             1   
10365        -1    3457.741260         63    BACHELOR'S DEGREE             0   
10918        -1    3404.990393         50  secondary education             1   
11109        -1    3438.463024         59  secondary education             1   
11268        -1    1048.782203         30    bachelor's degree             0   
11617        -1     370.827130         27  secondary education             1   
12304        -1    8493.101252         61         some college             2   
13786        -1    1900.579581         42  secondary education             1   
14357        -1     268.337037         23  secondary education             1   
15147        -1    9851.184337         44  secondary education             1   
15435        -1    3614.220232         40    bachelor's degree             0   
16129        -1     457.861760         33  secondary education             1   
16265        -1    2802.218127         40    bachelor's degree             0   
17095        -1    2809.693200         34  secondary education             1   
17429        -1     895.379738         37  secondary education             1   
17657        -1    4571.957475         41  secondary education             1   
18219        -1    3575.215641         33  secondary education             1   
19100        -1     617.246968         38  secondary education             1   
19366        -1     372.034749         43  SECONDARY EDUCATION             1   
19417        -1    3503.407602         28  secondary education             1   
20393        -1    3551.571072         69  secondary education             1   
20717        -1     661.822321         32    bachelor's degree             0   
21140        -1    1422.668059         44  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
291    civil partnership                 1      F       employee     0   
705              m

In [15]:
#Check children <0
data[data['children'] <0]['children'].value_counts()

#There are 47 rows with -1 children
#Maybe there was a mistake when submitting the values and the '-' sign is a mistake

-1    47
Name: children, dtype: int64

In [16]:
#Replace negative age to positive 
data.loc[data['children'] <0, 'children'] = data.loc[data['children'] <0, 'children'].abs()

In [17]:
#Check children = 20
data[data['children'] == 20].groupby('gender').count()
#does'nt look realistic for female to have 20 children (#metoo!!! :) ) 
#Maybe typo so 20 is actually  2 ->replace it 

#Abount men how have 20 children - can be realistic(#metoo chauvinist) more details in the conclusion part
 

children  days_employed  dob_years  education  education_id  \
gender                                                                
F             47             47         47         47            47   
M             29             29         29         29            29   

        family_status  family_status_id  income_type  debt  total_income  \
gender                                                                     
F                  47                47           47    47            47   
M                  29                29           29    29            29   

        purpose  
gender           
F            47  
M            29

In [18]:
# Female with 20 children -> change it to 2 
data.loc[(data['children'] == 20) & (data['gender'] == 'F'),'children'] = 2


### Conclusion

1. 'days_employed' data is not clear.
    There are positive and also negatine numbers. Is negative numbers are correct data?    
    transfering days to years gives un realistic values. What the values stand for? 
2. 'days_employed' column -taken an assumption that period od time that a person in working is 60 years.
    numbers above 21360(60 years) are divided by 10 until getting realistic number
3.  There are 15906 rows with negative days_employed. its look like a typo.
    negative days_employed was replaced to positive.
4. customers with total_income=NAN or total_income=0, there values were changed to the mean value of there income_type 
5. customers with days_employed=NAN or days_employed=0, there values were changed to the mean value of there income_type 
6. There are 101 rows that dob_years=0. maybe custumers didn't want to unveil there age.
    Becausee are not using the age as a parameter for our customer score i will not change/delete this column/rows
7. there is no relation between customers with age=0 to repaying a loan on time. a conclusion is that customers with debt          didn't afraid to unveil there age from a thougth that they will not get a loan
8. There are 47 rows with children=-1. I think its a mistake when submitting the numbers . the '-' sign is a mistake
   so i replaced negative numbers with positive number(-1 to 1)
9. does'nt look realistic for female to have 20 children (#metoo!!! :) ) 
   Maybe typo so 20 is actually  2 ->replace it 
   About men how have 20 children - can be realistic(#metoo!!!!!! )
   for example - moslem men with several wifes, Amish Mennonite...
   there are only 29 rows like this.
   I decide to keep it like this and not to change it to 2


### Data type replacement

In [19]:
#Check list of unique values, to make sure there is only valid values [0,1]
data.debt.unique()

#Replace 'debt' from object to boolean
data['debt'] = data['debt'].astype(bool)
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null bool
total_income        21525 non-null float64
purpose             21525 non-null object
dtypes: bool(1), float64(2), int64(4), object(5)
memory usage: 1.8+ MB


### Conclusion

1. 'debt' column values are [0,1] meaning that if a customer has ever defaulted on a loan. 
    Field can be changed to boolean

### Processing duplicates

In [20]:
#Get number of duplicated rows
data.duplicated().sum()

54

In [21]:
#Go through the duplicates and find conclusions 
data[data.duplicated()]

children  days_employed  dob_years            education  education_id  \
2849          0    2095.293025         41  secondary education             1   
4182          1    2095.293025         34    BACHELOR'S DEGREE             0   
4851          0    3259.094970         60  secondary education             1   
5557          0    3259.094970         58  secondary education             1   
7808          0    3259.094970         57  secondary education             1   
8583          0    3259.094970         58    bachelor's degree             0   
9238          2    2095.293025         34  secondary education             1   
9528          0    3259.094970         66  secondary education             1   
9627          0    3259.094970         56  secondary education             1   
10462         0    3259.094970         62  secondary education             1   
10697         0    1900.579581         40  secondary education             1   
10864         0    3259.094970         62  secondary education             1   
10994         0    3259.094970         62  secondary education             1   
11791         0    2095.293025         47  secondary education             1   
12373         0    2095.293025         58  secondary education             1   
12375         1    2095.293025         37  secondary education             1   
12736         0    3259.094970         59  secondary education             1   
13025         1    2095.293025         44  secondary education             1   
13639         0    3259.094970         64  secondary education             1   
13773         0    2095.293025         35  secondary education             1   
13878         1    1900.579581         31  secondary education             1   
13942         0    2095.293025         44  secondary education             1   
14432         2    3057.343890         36    bachelor's degree             0   
14832         0    2095.293025         50  secondary education             1   
15091         0    2095.293025         58  secondary education             1   
15188         0    2095.293025         60  secondary education             1   
15273         0    3259.094970         57  secondary education             1   
16176         0    2095.293025         47  secondary education             1   
16378         0    2095.293025         46  secondary education             1   
16902         2    2095.293025         39  secondary education             1   
16904         1    2095.293025         32    bachelor's degree             0   
17379         0    1900.579581         54    bachelor's degree             0   
17755         1    2095.293025         43  secondary education             1   
17774         1    1900.579581         40  secondary education             1   
18328         0    2095.293025         29    bachelor's degree             0   
18349         1    3057.343890         30    bachelor's degree             0   
18428         0    3259.094970         64  secondary education             1   
18521         0    2095.293025         56  secondary education             1   
18563         0    2095.293025         54  secondary education             1   
18755         0    3259.094970         58  secondary education             1   
19041         0    2095.293025         56  secondary education             1   
19184         0    2095.293025         46  secondary education             1   
19321         0    2095.293025         23  secondary education             1   
19387         0    1900.579581         38    bachelor's degree             0   
19688         0    3259.094970         61  secondary education             1   
19832         0    2095.293025         48  secondary education             1   
19946         0    2095.293025         57  secondary education             1   
20116         0    3259.094970         57  secondary education             1   
20165         0    2095.293025         42  secondary education             1   
20702  

In [22]:
#Check how many duplicated for each gender
data[data.duplicated()].groupby('gender')['gender'].count()

gender
F    47
M     7
Name: gender, dtype: int64

In [23]:
#Check total population for each gender
data.groupby('gender')['gender'].count()

gender
F      14236
M       7288
XNA        1
Name: gender, dtype: int64

In [24]:
#Drop duplicates
data.drop_duplicates(inplace=True)
#Varify there are no duplicates 
data.duplicated().sum()

0

### Conclusion

1. There are more duplicated rows for female then men.
     female = 47
     men = 7
2. The ratio of duplecates(54) to the total population is negligible 
   And also the ratio of duplicates by gender to the total population  is negligible
3. The 'days_employed' and 'debt' values of the duplicates is 0
4. Looks like the duplicates are a typo
5.  Conclusion - we can delete the duplicates rows

### Categorizing Data

In [25]:
import nltk
from nltk.stem import WordNetLemmatizer
#Working with lemmatzation
wordnet_lemma = WordNetLemmatizer() 

In [26]:
#Diagnosis the values of 'purpose' column
data['purpose'].value_counts()

wedding ceremony                            793
having a wedding                            773
to have a wedding                           769
real estate transactions                    675
buy commercial real estate                  662
housing transactions                        652
buying property for renting out             652
transactions with commercial real estate    650
purchase of the house                       646
housing                                     646
purchase of the house for my family         638
construction of own property                635
property                                    633
transactions with my real estate            627
building a real estate                      625
buy real estate                             621
purchase of my own house                    620
building a property                         619
housing renovation                          607
buy residential real estate                 606
buying my own car                       

In [27]:
# find consecutive words
#example ['buy', 'commercial', 'real', 'estate'] will return ['buy', 'commercial', 'real estate']
def find_consecutive_words(lemmas_list, list_to_check):
    for i, word in enumerate(list_to_check):
         if i+1 < len(list_to_check):
            word_pair = word + ' '+ list_to_check[i+1]   #Get pair of words         
            if word_pair in lemmas_list: #Check if  exist in lemmas_list
                del list_to_check[i+1]   #Delete word  for example 'real'
                del list_to_check[i]     #Delete word  foer example 'state'
                list_to_check.insert(i,word_pair) # Add onsecutive word 'real estate' 
    return list_to_check
#create purpose categories 
def get_purpose_category(purpose,real_estate_lemmas,education_list_lemmas,car_lemma,wedding_lemma):
    
    purpose_words = nltk.word_tokenize(purpose)
    purpose_words_lemmas = [wordnet_lemma.lemmatize(word,pos="n") for word in purpose_words]
    purpose_words_lemmas.extend([wordnet_lemma.lemmatize(word,pos="v") for word in purpose_words])
     
    #  find consecutive words   
    purpose_words_lemmas = find_consecutive_words(real_estate_lemmas, purpose_words_lemmas)
    #get unique values/distinct values
    purpose_words_lemmas = set(purpose_words_lemmas)

    purpose_category = None
    for word in purpose_words_lemmas:     
        if word in real_estate_lemmas:
            purpose_category = 'real estate'
            break;
        elif word in education_list_lemmas:
            purpose_category = 'education'
            break;
        elif word in car_lemma:
            purpose_category = 'car'
            break;
        elif word in wedding_lemma:
            purpose_category = 'wedding'
            break;       
    return purpose_category

#real estate category
real_estate_list = ['house', 'property', 'real estate']
real_estate_lemmas = [wordnet_lemma.lemmatize(x,pos="n") for x in real_estate_list]
real_estate_lemmas.extend([wordnet_lemma.lemmatize(x,pos="v") for x in real_estate_list])
#get unique values/distinct values
real_estate_lemmas = set(real_estate_lemmas)

#education category
education_list = ['education', 'university','educate','educated']
education_list_lemmas =  [wordnet_lemma.lemmatize(x,pos="n") for x in education_list]
education_list_lemmas.extend([wordnet_lemma.lemmatize(x,pos="v") for x in education_list])

#get unique values/distinct values
education_list_lemmas = set(education_list_lemmas)

#car category
car_lemma = []
car_lemma.append(wordnet_lemma.lemmatize('car',pos="n"))
car_lemma.append(wordnet_lemma.lemmatize('car',pos="v"))

#wedding category
wedding_lemma = []
wedding_lemma.append(wordnet_lemma.lemmatize('wedding',pos="n"))
wedding_lemma.append(wordnet_lemma.lemmatize('wedding',pos="v"))

#create 'purpose_category' column
data['purpose_category'] = data['purpose'].apply(get_purpose_category,args=(real_estate_lemmas,education_list_lemmas,car_lemma,wedding_lemma ))
data


children  days_employed  dob_years            education  education_id  \
0             1    8437.673028         42    bachelor's degree             0   
1             1    4024.803754         36  secondary education             1   
2             0    5623.422610         33  Secondary Education             1   
3             3    4124.747207         32  secondary education             1   
4             0    3402.660720         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21520         1    4529.316663         43  secondary education             1   
21521         0    3439.374041         67  secondary education             1   
21522         1    2113.346888         38  secondary education             1   
21523         3    3112.481705         38  secondary education             1   
21524         2    1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type   debt  \
0                married                 0      F    employee  False   
1                married                 0      F    employee  False   
2                married                 0      M    employee  False   
3                married                 0      M    employee  False   
4      civil partnership                 1      F     retiree  False   
...                  ...               ...    ...         ...    ...   
21520  civil partnership                 1      F    business  False   
21521            married                 0      F     retiree  False   
21522  civil partnership                 1      M    employee   True   
21523            married                 0      M    employee   True   
21524            married                 0      F    employee  False   

       total_income                  purpose purpose_category  
0         40620.102    purchase of the house      real estate  
1         17932.802             car purchase              car  
2         23341.752    purchase of the house      real estate  
3         42820.568  supplementary education        education  
4         25378.572        to have a wedding          wedding  
...             ...                      ...              ...  
21520     35966.698     housing transactions      real estate  
21521     24959.969        purchase of a car              car  
21522     14347.610                 property      real estate  
21523     39054.888        buying my own car              car  
21524     13127.587             to buy a car              car  

[21471 rows x 13 columns]

######   Categoriz 'total_income' column

In [29]:
data.describe()


children  days_employed     dob_years  education_id  \
count  21471.000000   21471.000000  21471.000000  21471.000000   
mean       0.504541    2558.881781     43.279074      0.817195   
std        1.040344    2048.105302     12.574291      0.548508   
min        0.000000      24.141633      0.000000      0.000000   
25%        0.000000    1024.119744     33.000000      1.000000   
50%        0.000000    2095.293025     42.000000      1.000000   
75%        1.000000    3568.078677     53.000000      1.000000   
max       20.000000   18388.949901     75.000000      4.000000   

       family_status_id   total_income  
count      21471.000000   21471.000000  
mean           0.973685   26787.530220  
std            1.421082   15683.865445  
min            0.000000    3306.762000  
25%            0.000000   17224.844000  
50%            0.000000   24303.791000  
75%            1.000000   32386.793835  
max            4.000000  362496.645000

In [30]:
#return the  total_income_category for each total_income
def get_total_income_category(total_income):
    total_income_category = None
    
    if total_income<16488.504500:
        total_income_category = 'low'
    elif 16488.504500 <total_income< 23202.870000:
        total_income_category = 'medium'
    elif 23202.870000<total_income<  32549.611000:
        total_income_category = 'high'
    elif total_income>  32549.611000:
        total_income_category = 'super high'
    return total_income_category

data['total_income_category'] = data['total_income'].apply(get_total_income_category)


### Conclusion

1. There are 4 main types of  loans:
    real estate
    wedding
    education
    car
    
2. The categories and there values  will be:
    real estate - house, property, real estate
    wedding - wedding(there are no Synonyms)
    education - education, university,educate,educated
    car - car (there are no Synonyms)
3. build 4 categories for 'total_income'
    1. low  = total_income<16488.504500    25% of  customers salary
    2. medium = 16488.504500 <total_income<23202.870000  25% - 50% of  customers salary
    3 high = 23202.870000<total_income<  32549.611000     50% - 75% of  customers salary
    4 super high = total_income>  32549.611000   75% of  customers salary

## Answer these questions

- Is there a relation between having kids and repaying a loan on time?

In [31]:
#use pivot_table
data_pivot = data.pivot_table(
    index='children',
    columns='debt',
    values='education_id',
    aggfunc='count',
)
#Replace NaN with 0
data_pivot.fillna(0, inplace=True)
#Check if we take care of all NaN 
data_pivot.isnull().sum()

#Calculate 'ratio' column to see if there is a relation between having kids and repaying a loan on time 
data_pivot['ratio'] = data_pivot[1] / data_pivot[0]

data_pivot

debt        False    True     ratio
children                           
0         13044.0  1063.0  0.081493
1          4411.0   445.0  0.100884
2          1901.0   198.0  0.104156
3           303.0    27.0  0.089109
4            37.0     4.0  0.108108
5             9.0     0.0  0.000000
20           25.0     4.0  0.160000

### Conclusion

1. As we can see by the 'ratio' column, the increcment of children  affecting repaying a loan on time

- Is there a relation between marital status and repaying a loan on time?

In [32]:
# pivot_table
data_pivot = data.pivot_table(
    index='family_status',
    columns='debt',
    values='education_id',
    aggfunc='count',
)

#Calculate 'ratio' column to see if there is a relation between  marital status and repaying a loan on time 
data_pivot['ratio'] = data_pivot[1] / data_pivot[0]
data_pivot

debt               False  True     ratio
family_status                           
civil partnership   3775   388  0.102781
divorced            1110    85  0.076577
married            11413   931  0.081574
unmarried           2536   274  0.108044
widow / widower      896    63  0.070312

### Conclusion

 As we can see by the 'ratio' column, castumer with civil partnership and unmarried marital are more risky than the others 

- Is there a relation between income level and repaying a loan on time?

In [33]:
# pivot_table
data_pivot = data.pivot_table(
    index='total_income_category',
    columns='debt',
    values='education_id',
    aggfunc='count',
)

#Calculate 'ratio' column to see if there is a relation between income level and repaying a loan on time 
data_pivot['ratio'] = data_pivot[1] / data_pivot[0]
data_pivot


debt                   False  True     ratio
total_income_category                       
high                    6001   561  0.093484
low                     4455   383  0.085971
medium                  4775   456  0.095497
super high              4498   341  0.075811

### Conclusion

As we can see by the 'ratio' column, total_income doesn't affect in repaying a loan on time

- How do different loan purposes affect on-time repayment of the loan?

In [34]:
# pivot_table
data_pivot = data.pivot_table(
    index='purpose_category',
    columns='debt',
    values='education_id',
    aggfunc='count',
)

#Calculate 'ratio' column to see if there is a relation between loan purposes and repaying a loan on time 
data_pivot['ratio'] = data_pivot[1] / data_pivot[0]
data_pivot


debt              False  True     ratio
purpose_category                       
car                3905   403  0.103201
education          3644   370  0.101537
real estate       10032   782  0.077951
wedding            2149   186  0.086552

### Conclusion

car loan and education loan are more risky then real estate and wedding loans

## General conclusion

In [35]:
#pivot table
data_pivot = data.pivot_table(
    index=['total_income_category','family_status'],
    columns='debt',
    values='education_id',
    aggfunc='count',
)

#Calculate 'ratio' column to see if there is a relation between income level and repaying a loan on time 
data_pivot['ratio'] = data_pivot[1] / data_pivot[0]
data_pivot.head(50)

debt                                     False  True     ratio
total_income_category family_status                           
high                  civil partnership   1143   138  0.120735
                      divorced             345    25  0.072464
                      married             3520   284  0.080682
                      unmarried            783    93  0.118774
                      widow / widower      210    21  0.100000
low                   civil partnership    857    87  0.101517
                      divorced             243    21  0.086420
                      married             2536   212  0.083596
                      unmarried            529    49  0.092628
                      widow / widower      290    14  0.048276
medium                civil partnership    903   102  0.112957
                      divorced             263    22  0.083650
                      married             2699   238  0.088181
                      unmarried            654    74  0.113150
                      widow / widower      256    20  0.078125
super high            civil partnership    872    61  0.069954
                      divorced             259    17  0.065637
                      married             2657   197  0.074144
                      unmarried            570    58  0.101754
                      widow / widower      140     8  0.057143

Customers  that has super high income and thay are NOT unmarried have the best credit score  